### DISCUSSION TITLE
Episode: Introduction and setup

### DISCUSSION THREAD

**Discussion by kevinrue** [carpentries-incubator/bioc-project#36] [Ideas] [answered]
The list below is a suggestion of individual items in this episode.

Feedback is welcome!

Ticked boxes indicate items that already have a dedicated section in this episode.

Topics

- [x] Correct R version is key to access correct Bioconductor version

Challenges

- [x] `R.version.string`

Notes

- Does the patch version influence whether `BiocManager::valid()` detects out of date base R packages?

**Comment by kevinrue** [comment, marked as answer]
This episode is mostly done at this point: https://carpentries-incubator.github.io/bioc-project/01-setup/index.html

**Comment by doujouDC** [comment]
Hi, sorry I am new to requesting changes on Git. So, hopefully this is the right place. In the setup rmd lines 47-53 the code appears to be creating the same file twice. I think the last line should be creating this file: "actb_orfs.fasta"? Also, I am planning on using these lessons as part of an Introduction to Bioconductor class I am developing. Where is the most appropriate place to provide feedback?

Thanks so much for putting these together

**Comment by kevinrue** [comment]
Well spotted, thanks!

I'm just adding the link to the mistake here for reference: https://github.com/carpentries-incubator/bioc-project/blob/main/_episodes_rmd/01-setup.Rmd#L54

If you're new to git (the program) and github (the website), then there's a bit of learning curve ahead of you. Not a crazy amount, but a bit too much to detail it all here.

For now, allow me to skip best practices and suggest the following:

- Click on this link (https://github.com/carpentries-incubator/bioc-project/blob/main/_episodes_rmd/01-setup.Rmd) to navigate to the file that needs to be edited (you could navigate to it yourself as you clearly did already)
- At the top of the "file" on the right click on the pencil icon

<img width="312" alt="image" src="https://user-images.githubusercontent.com/5769108/216121137-bd6876bd-c0a6-4272-8aa3-7a1333cfb999.png">

- On line 54, replace `"data/actb.gtf"` by `"data/actb_orfs.gtf"`
- At the bottom of the page, make sure that "create a branch" is selected (the name of the branch doesn't matter much, I've screenshot'ed the default one that was offered to me)

<img width="705" alt="image" src="https://user-images.githubusercontent.com/5769108/216121746-51ae9c69-344b-43d1-bf14-56c0774b9fad.png">

- Click on "Propose changes"
- On the next page, click again on the green button to open a "pull request", which will give me the possibility to accept your change into the main repository.

Any issue, let me know!

PS: It's a simple change that I could do myself, now that you pointed it out, but you seem keen to learn about git and github, so I thought I'd give you a little taste of what it feels like to work with git and github. Plus, doing it this way will get your name into the history of contributors, which is always a nice acknowledgment of your contribution :)

**Reply by doujouDC** [reply]
Kevin,

My interface looked different than yours so I did not want to change anything. This is what I saw at the bottom of the page (after I made the edits)
<img width="1320" alt="Screen Shot 2023-02-01 at 14 52 27" src="https://user-images.githubusercontent.com/6465940/216149180-e3008d72-381b-44de-a2f5-ae6c9780a3be.png">

So, no options to create a new branch. I did however, notice this at the top of the page:
<img width="1335" alt="Screen Shot 2023-02-01 at 14 51 54" src="https://user-images.githubusercontent.com/6465940/216149270-a5e601b6-7f3c-4f48-9579-7d0b27965bc9.png">

**Comment by doujouDC** [comment]
Hi Kevin,

Thanks so much for the tips. I think I will try to make the change myself
if that’s OK  with you.

**Comment by kevinrue** [comment]
Yes that's alright go ahead. You can't break anything on the main repo.
It's saying you're about to make a working copy of it on your account
instead. All good!

**Reply by doujouDC** [reply]
Thanks so much much, I am not sure if I "pulled" the first request. I think I might have found a couple of possible spelling errors. For those I think I did do a pull request

Doug

**Reply by kevinrue** [reply]
Yeah I don't see the first one that we discussed, but I see another two fixing typos. I'll merge those while you try and re-do the first one again, ok?

**Comment by doujouDC** [comment]
Thanks again. The pull request should have gone through

Since I plan on teaching the first two lessons that you developed in May I
did have a couple of questions. In this image, the "?' are because you are
not sure of the date, or you need someone to fill in the dates?

[image: bioc-timeline.png]

This is image, the bottom four boxes are blank is this intentional, or to
signify that there are lots more items under BAM

[image: bioc-sequencing-ecosystem.png]

Last, what type of attribution would you like?

Thanks so much for your help and patience

Sent from my iPhone

On Feb 2, 2023, at 04:49, Kevin Rue-Albrecht ***@***.***>
wrote:

﻿

Yeah I don't see the first one that we discussed, but I see another two
fixing typos. I'll merge those while you try and re-do the first one again,
ok?

—
Reply to this email directly, view it on GitHub
<https://github.com/carpentries-incubator/bioc-project/discussions/36#discussioncomment-4849759>,
or unsubscribe
<https://github.com/notifications/unsubscribe-auth/ABRKTFGTGYZURRR2Z2WAYVTWVN7KPANCNFSM5HHRBCKQ>
.
You are receiving this because you commented.Message ID:
<carpentries-incubator/bioc-project/repo-discussions/36/comments/4849759@
github.com>

**Reply by kevinrue** [reply]
Thanks for your thorough review!

> In this image, the "?' are because you are
not sure of the date, or you need someone to fill in the dates?

I put ? as placeholder hoping some of the earliest Bioconductor community members could fill in the pieces I couldn't dig up from the web.

> This is image, the bottom four boxes are blank is this intentional, or to signify that there are lots more items under BAM

Here, I meant the blank boxes to signify that there are lots more items possible, and that the ones I showed above are only examples.

> Last, what type of attribution would you like?

I don't understand what you mean here.

**Reply by doujouDC** [reply]
Hi Kevin, sorry if that last comment was out of context. I plan on adapting the first two lessons for a class that I am teaching. I wanted to identify that the content was modified/based on your repo. So, I wanted to know how you wanted me to cite you. Does this help?

Doug

**Reply by kevinrue** [reply]
First, thanks for asking :)

I'm not entirely sure what to recommend. You'll see that the license is fairly relaxed (https://carpentries-incubator.github.io/bioc-project/LICENSE.html)

In particular:

> Attribution—You must give appropriate credit (mentioning that your work is derived from work that is Copyright © Software Carpentry and, where practical, linking to http://software-carpentry.org/), provide a [link to the license](https://creativecommons.org/licenses/by/4.0/), and indicate if changes were made. You may do so in any reasonable manner, but not in any way that suggests the licensor endorses you or your use.

I also realise now that I haven't even bothered putting my name in the authors file (https://github.com/carpentries-incubator/bioc-project/blob/main/AUTHORS), which is what is linked at the bottom of every page in the lesson.

I'll tell you what I usually do, as I regularly give a workshop about Git and GitHub, which is partly inspired from The Carpentries lesson "[Version Control with Git](https://swcarpentry.github.io/git-novice/index.html)". I have a slide toward the end of my presentation, dedicated to a large logo of The Carpentries and a link to the lesson. That's it. I always verbally encourage my students to go check out the lesson themselves after the workshop, as they may learn something more/differently from how I presented it myself. 

Bottom line is that you don't need to credit me personally for this lesson. I've populated most of the contents so far myself, yes, but the lesson is bigger than me, open for anyone to contribute, and I also credit other people for setting up the lesson template and automated testing/deployment, without which I wouldn't have been able to fill the contents in the first place :)

**Reply by doujouDC** [reply]
Thanks, I usually do that too, for Carpentries approved lessons. I was just asking since the incubator lessons are not really sanctioned by the Carpentries. I just wanted to make sure I cite someone. Thanks so much for the feedback.

Cheers

Doug

**Reply by kevinrue** [reply]
Good point. Feel free to use my name, but you probably also want to acknowledge the Bioconductor Education 
 working group, which I'm part of, but leads the lesson development as a group: http://workinggroups.bioconductor.org/currently-active-working-groups-committees.html#education

**Reply by tobyhodges** [reply]
Thank you both for the thoughtful approach you are taking to this question. @kevinrue I can recommend that you make a Zenodo publication from the lesson repository (and the other Bioconductor lessons), which you can then update with milestone releases (e.g. first alpha version, first beta version, etc) as development continues. you could add a `.zenodo.json` file to the repository to save time with populating the Zenodo entry, basing the author list on the list of contribtutors for the repository. Finally, you could update the `CITATION` file in the repository with the DOI you get from Zenodo.

**Comment by doujouDC** [comment]
Hi, I have a quesiton about the biocView challenge. In the challenge the lesson says "Visit the listing of all packages on the Bioconductor [biocViews](https://www.bioconductor.org/packages/release/BiocViews.html) web page. Use the “Autocomplete biocViews search” box in the upper left to filter packages by category and explore the graph of software packages by expanding and contracting individual terms."

It is not clear to me what this means. Are you suggesting that the learner use the search box to search for one of the terms list in the challenge. For example "RNA sequencing analysis (RNASeq" If so, I do not see a "graph" for explore. Below is a screenshot that results from a search of "RNASeq"

<img width="960" alt="Screen Shot 2023-02-06 at 13 52 11" src="https://user-images.githubusercontent.com/6465940/217060194-acfd66ed-38b9-424f-a100-49c2e1a4be9a.png">

So the goal of this challenge is identifying the 369 packages that are related to "RNAseq"

Also for the solution you highlight DESeq2 and edgeR as being the most frequently downloaded packages. However, it you sort by number of downloads in the table, these are the top downloads:

XCIR | 2183
MetaPhOR  | 2172
ISLET | 2158
...

Thanks

Doug

**Reply by kevinrue** [reply]
I've picked up the word "graph" from elsewhere in Bioconductor teaching materials, but I agree that I wouldn't call it a "graph" either. I'm not sure what to call it, but I'm talking about the nested hierarchy of "tags" that the community calls "BiocViews".

I'm also not entirely sure how the download stats are converted into the rank metric. Downloads are counted monthly. Perhaps the rank refers to the total number of downloads since each package was added (seems unfair, but I really don't know how the rank is computed).

What I'm taking from your comment is that the actual expectations of the challenge should be clarified.
My intention here was simply to get participants to play with this page, navigate the classification of packages, and identify the most popular packages for each task (I always point out that package that have a bigger user base are generally easier to find help and advice, in particular in the Bioc support site).

**Reply by doujouDC** [reply]
Hi Kevin!

Thanks for taking the time to reply as I know you are busy with teaching.  This is the way that I was handling this section for the Intro class that I adapted from the material. The section below reflect my teaching style, so it is only a suggestion. Please don't feel like to need to modify the repo. As you can see, I modified some of the other sections too. 

# Bioconductor Packages

## Software Packages

[Software packages](https://carpentries-incubator.github.io/bioc-project/reference.html#software-package) can be subdivided into packages that provide infrastructure (i.e., classes) to store and access data, and packages that provide methodological tools to process data stored in those data structures.

This separation of structure and analysis is at the core of the Bioconductor project, encouraging developers of new methodological software packages to thoughtfully re-use existing data containers where possible, and reducing the cognitive burden imposed on users who can more easily experiment with alternative workflows without the need to learn and convert between different data structures.

## Annotation Packages

[Annotation data packages](https://carpentries-incubator.github.io/bioc-project/reference.html#annotationdata-package) provide self-contained databases of diverse genomic annotations (e.g., gene identifiers, biological pathways).

Annotation packages are identifiable by their respective naming pattern, and the information that they contain. For instance, the so-called `OrgDb` packages (e.g., the [*org.Hs.eg.db*](https://bioconductor.org/packages/3.16/org.Hs.eg.db) package) provide information mapping different types of gene identifiers and pathway databases. Another example is the `TxDb` packages (e.g., [*TxDb.Hsapiens.UCSC.hg38.knownGene*](https://bioconductor.org/packages/3.16/TxDb.Hsapiens.UCSC.hg38.knownGene)) which encapsulate individual versions UCSC gene annotation tables.

## Experimental Data Packages

[Experiment data packages](https://carpentries-incubator.github.io/bioc-project/reference.html#experimentdata-package) provide self-contained datasets that are often used by software package developers to demonstrate the use of their package on well-known standard datasets in their [package vignettes](https://carpentries-incubator.github.io/bioc-project/reference.html#vignette).

## Workflow Packages

[Workflow packages](https://carpentries-incubator.github.io/bioc-project/reference.html#workflow-package) exclusively provide collections of vignettes that demonstrate the combined usage of several other packages as a coherent workflow, but do not provide any new source code or functionality themselves.

## Challenge 1: Exploring the Bioconductor website

Open up the Bioconductor [website](https://bioconductor.org/)

Browse the website to find information answering the following questions:

1.  How many packages does the current release of the Bioconductor project include?

2.  How many packages of each type does this number include?

### Challenge 1: Answer

The following solution includes numbers that were valid at the time of writing (Bioconductor release 3.16); numbers will inevitably be different for future releases of the Bioconductor project.

From the Bioconductor [homepage](https://www.bioconductor.org/):

1.  From the Install section, you can read "Discover 2183 software packages available in Bioconductor release 3.16"
2.  From the "News" section, click on the link that reads Bioconductor [Bioc 3.16](https://bioconductor.org/news/bioc_3_16_release) is the current release.

# **Package Classification using biocViews**

Bioconductor uses [biocViews](https://carpentries-incubator.github.io/bioc-project/reference.html#biocviews), a set of terms from a controlled vocabulary, to classify Bioconductor packages and facilitate their discovery by thematic search on the [Bioconductor website](https://www.bioconductor.org/packages/release/BiocViews.html).

Each Bioconductor package is tagged with a small set of terms chosen from the available controlled vocabulary. These terms describe the type and functionality of the package. Terms are initially selected by the package authors, and subsequently refined during package review or updates to the controlled vocabulary.

## Challenge 2: Exploring biocViews

Visit the listing of all packages on the Bioconductor [biocViews](https://www.bioconductor.org/packages/release/BiocViews.html) web page. You can use the "Autocomplete biocViews search" box in the upper left to filter packages by category.

1.  What biocView terms can be used to identify packages that have been tagged for:

    -   RNA sequencing analysis?

    -   ChIP-seq?

    -   Epigenetics?

2.  In the `RNASeq` category, two very popular packages are [*DESeq2*](https://bioconductor.org/packages/3.16/DESeq2) and [*edgeR*](https://bioconductor.org/packages/3.16/bioc/html/DESeq2.html). Which one is more popular in terms of download statistics (i.e., higher rank)?

### Challenge 2: Exploring biocViews Answer

1.  `RNAseq`, `ChIPSeq`, and `Epigenetics`.

2.  For Bioconductor release `3.16`, [*DESeq2*](https://bioconductor.org/packages/3.16/DESeq2) and [*edgeR*](https://bioconductor.org/packages/3.16/edgeR) are listed at ranks 29 and 23 respectively. In other words, the two packages are among the most frequently downloaded packages in the Bioconductor project, in this instance with a small advantage in favour of [*edgeR*](https://bioconductor.org/packages/3.16/edgeR).

# Resources

-   Package [vignettes](https://www.bioconductor.org/help/package-vignettes/) and manuals: each Bioconductor package contains at least one vignette, a document that provides a task-oriented description of package functionality. Vignettes contain executable examples and are intended to be used interactively. You can access the PDF version of a vignette for any installed package from inside R using the `browseVignettes` function.

-   [Workflows](https://www.bioconductor.org/packages/release/workflows/) for learning and use. Workflows vignettes are now available as packages. View all [current workflows](https://www.bioconductor.org/packages/release/BiocViews.html#___Workflow).

-   Several [online books](https://www.bioconductor.org/books/release/) for comprehensive coverage of a particular research field, biological question, or technology.

-   [Course and conference](https://www.bioconductor.org/help/course-materials/) material.

-   [Videos](https://www.youtube.com/user/bioconductor).

-   Community [resources](https://www.bioconductor.org/help/community/) and [tutorials](https://support.bioconductor.org/t/Tutorials/).

# Licenses

Licensed under [CC-BY 4.0](https://creativecommons.org/licenses/by/4.0/) Add any licensing information or sources used in the creation of this class. This class was adapted from lessons developed by the Carpentries [Incubator Bioc-Project](https://github.com/carpentries-incubator/bioc-project)
